AWSAppSyncClient/Internal/AppSyncSubscriptionWithSync.swift

```diff
@@ -0,0 +1,596 @@
+            hasDeltaQuery: self.deltaQuery != nil,
+            syncRefreshIntervalInSeconds: syncConfiguration.syncIntervalInSeconds
+        )
+
+        internalStateSyncQueue = OperationQueue()
+        internalStateSyncQueue.maxConcurrentOperationCount = 1
+        internalStateSyncQueue.name = "AppSync.DeltaSyncOperationQueue.\(getOperationHash())"
+
+        performInitialSync()
+    }
+
+    // MARK: - Cancellable
+
+    /// Cancel the subscription sync. This releases all internal resources and unregisters for system notifications.
+    ///
+    /// Callers must invoke `cancel` on subscription sync in order to release resources. Failure to do so will cause
+    /// memory leaks.
+    func cancel() {
+        // perform user-cancellation tasks, if any, then internal ones.
+        internalCancel()
+    }
+
+    // MARK: - Setup
+
+    private func performInitialSync() {
+        AppSyncLog.debug("Queuing operations for initial sync")
+        internalStateSyncQueue.addOperation {
+            self.registerForNotifications()
+            self.initializeLastSyncTimeFromCache()
+            self.initializeBaseQueryResultsFromCache()
+            self.performSync()
+        }
+    }
+
+    // MARK: - Sync time
+
+    /// Fetches last sync time from local cache and updates `syncStrategy` with the retrieved value
+    private func initializeLastSyncTimeFromCache() {
+        do {
+            let lastSyncTime = try self.subscriptionMetadataCache?.getLastSyncTime(operationHash: getOperationHash())
+            syncStrategy.lastSyncTime = lastSyncTime
+            AppSyncLog.debug("Loaded lastSyncTime from cache: \(lastSyncTime.debugDescription)")
+        } catch {
+            // could not find it in cache, assume no sync was done previously
+        }
+    }
+
+    /// Updates the last sync time and persists it to `subscriptionMetadataCache` if present. Expected to
+    /// be called when subscription message is given to the caller, or if base query or delta query is run.
+    ///
+    /// - Parameters:
+    ///   - date: The date at which the last sync was successfully performed
+    private func setLastSyncTime(date: Date) {
+        do {
+            let adjustedDate = date.addingTimeInterval(TimeInterval.init(exactly: -2)!)
+            AppSyncLog.debug("Updating lastSync time \(adjustedDate)")
+            self.syncStrategy.lastSyncTime = adjustedDate
+            try subscriptionMetadataCache?.updateLastSyncTime(for: getOperationHash(), with: adjustedDate)
+        } catch {
+            // ignore cache write failure, be updated in next operation, is backed up by in-memory cache
+        }
+    }
+
+    // MARK: - Base query handling
+
+    /// Load base query results from cache, to quickly provide callers with the most recent results
+    private func initializeBaseQueryResultsFromCache() {
+        guard let appSyncClient = appSyncClient else {
+            return
+        }
+
+        let dispatchGroup = DispatchGroup()
```

Comment:
I think that `DispatchGroup` has slightly different purpose, i.e. when you need to sync several async operations, but here you have only one async operation. So DispatchSemaphore suits better for this case.

Reply:
This might be a good suggestion to also align with other similar implementation in our sync implementation

Reply:
Thank you. This's I forgot to mention.

Reply:
Thanks, I neglected to replace this DispatchGroup with a semaphore, which I did elsewhere. Will change.

Reply:
(However, I'm not going to refactor all the other usages of DispatchGroup as part of this PR.)